// sucrase/dist/esm/parser/tokenizer/types.js
// Generated file, do not edit! Run "yarn generate" to re-generate this file.
/* istanbul ignore file */
/**
 * Enum of all token types, with bit fields to signify meaningful properties.
 */
export var TokenType; (function (TokenType) {
  // Precedence 0 means not an operator; otherwise it is a positive number up to 12.
  const PRECEDENCE_MASK = 0xf; TokenType[TokenType["PRECEDENCE_MASK"] = PRECEDENCE_MASK] = "PRECEDENCE_MASK";
  const IS_KEYWORD = 1 << 4; TokenType[TokenType["IS_KEYWORD"] = IS_KEYWORD] = "IS_KEYWORD";
  const IS_ASSIGN = 1 << 5; TokenType[TokenType["IS_ASSIGN"] = IS_ASSIGN] = "IS_ASSIGN";
  const IS_RIGHT_ASSOCIATIVE = 1 << 6; TokenType[TokenType["IS_RIGHT_ASSOCIATIVE"] = IS_RIGHT_ASSOCIATIVE] = "IS_RIGHT_ASSOCIATIVE";
  const IS_PREFIX = 1 << 7; TokenType[TokenType["IS_PREFIX"] = IS_PREFIX] = "IS_PREFIX";
  const IS_POSTFIX = 1 << 8; TokenType[TokenType["IS_POSTFIX"] = IS_POSTFIX] = "IS_POSTFIX";
  const IS_EXPRESSION_START = 1 << 9; TokenType[TokenType["IS_EXPRESSION_START"] = IS_EXPRESSION_START] = "IS_EXPRESSION_START";

  const num = 512; TokenType[TokenType["num"] = num] = "num"; // num startsExpr
  const bigint = 1536; TokenType[TokenType["bigint"] = bigint] = "bigint"; // bigint startsExpr
  const decimal = 2560; TokenType[TokenType["decimal"] = decimal] = "decimal"; // decimal startsExpr
  const regexp = 3584; TokenType[TokenType["regexp"] = regexp] = "regexp"; // regexp startsExpr
  const string = 4608; TokenType[TokenType["string"] = string] = "string"; // string startsExpr
  const name = 5632; TokenType[TokenType["name"] = name] = "name"; // name startsExpr
  const eof = 6144; TokenType[TokenType["eof"] = eof] = "eof"; // eof
  const bracketL = 7680; TokenType[TokenType["bracketL"] = bracketL] = "bracketL"; // [ startsExpr
  const bracketR = 8192; TokenType[TokenType["bracketR"] = bracketR] = "bracketR"; // ]
  const braceL = 9728; TokenType[TokenType["braceL"] = braceL] = "braceL"; // { startsExpr
  const braceBarL = 10752; TokenType[TokenType["braceBarL"] = braceBarL] = "braceBarL"; // {| startsExpr
  const braceR = 11264; TokenType[TokenType["braceR"] = braceR] = "braceR"; // }
  const braceBarR = 12288; TokenType[TokenType["braceBarR"] = braceBarR] = "braceBarR"; // |}
  const parenL = 13824; TokenType[TokenType["parenL"] = parenL] = "parenL"; // ( startsExpr
  const parenR = 14336; TokenType[TokenType["parenR"] = parenR] = "parenR"; // )
  const comma = 15360; TokenType[TokenType["comma"] = comma] = "comma"; // ,
  const semi = 16384; TokenType[TokenType["semi"] = semi] = "semi"; // ;
  const colon = 17408; TokenType[TokenType["colon"] = colon] = "colon"; // :
  const doubleColon = 18432; TokenType[TokenType["doubleColon"] = doubleColon] = "doubleColon"; // ::
  const dot = 19456; TokenType[TokenType["dot"] = dot] = "dot"; // .
  const question = 20480; TokenType[TokenType["question"] = question] = "question"; // ?
  const questionDot = 21504; TokenType[TokenType["questionDot"] = questionDot] = "questionDot"; // ?.
  const arrow = 22528; TokenType[TokenType["arrow"] = arrow] = "arrow"; // =>
  const template = 23552; TokenType[TokenType["template"] = template] = "template"; // template
  const ellipsis = 24576; TokenType[TokenType["ellipsis"] = ellipsis] = "ellipsis"; // ...
  const backQuote = 25600; TokenType[TokenType["backQuote"] = backQuote] = "backQuote"; // `
  const dollarBraceL = 27136; TokenType[TokenType["dollarBraceL"] = dollarBraceL] = "dollarBraceL"; // ${ startsExpr
  const at = 27648; TokenType[TokenType["at"] = at] = "at"; // @
  const hash = 29184; TokenType[TokenType["hash"] = hash] = "hash"; // # startsExpr
  const eq = 29728; TokenType[TokenType["eq"] = eq] = "eq"; // = isAssign
  const assign = 30752; TokenType[TokenType["assign"] = assign] = "assign"; // _= isAssign
  const preIncDec = 32640; TokenType[TokenType["preIncDec"] = preIncDec] = "preIncDec"; // ++/-- prefix postfix startsExpr
  const postIncDec = 33664; TokenType[TokenType["postIncDec"] = postIncDec] = "postIncDec"; // ++/-- prefix postfix startsExpr
  const bang = 34432; TokenType[TokenType["bang"] = bang] = "bang"; // ! prefix startsExpr
  const tilde = 35456; TokenType[TokenType["tilde"] = tilde] = "tilde"; // ~ prefix startsExpr
  const pipeline = 35841; TokenType[TokenType["pipeline"] = pipeline] = "pipeline"; // |> prec:1
  const nullishCoalescing = 36866; TokenType[TokenType["nullishCoalescing"] = nullishCoalescing] = "nullishCoalescing"; // ?? prec:2
  const logicalOR = 37890; TokenType[TokenType["logicalOR"] = logicalOR] = "logicalOR"; // || prec:2
  const logicalAND = 38915; TokenType[TokenType["logicalAND"] = logicalAND] = "logicalAND"; // && prec:3
  const bitwiseOR = 39940; TokenType[TokenType["bitwiseOR"] = bitwiseOR] = "bitwiseOR"; // | prec:4
  const bitwiseXOR = 40965; TokenType[TokenType["bitwiseXOR"] = bitwiseXOR] = "bitwiseXOR"; // ^ prec:5
  const bitwiseAND = 41990; TokenType[TokenType["bitwiseAND"] = bitwiseAND] = "bitwiseAND"; // & prec:6
  const equality = 43015; TokenType[TokenType["equality"] = equality] = "equality"; // ==/!= prec:7
  const lessThan = 44040; TokenType[TokenType["lessThan"] = lessThan] = "lessThan"; // < prec:8
  const greaterThan = 45064; TokenType[TokenType["greaterThan"] = greaterThan] = "greaterThan"; // > prec:8
  const relationalOrEqual = 46088; TokenType[TokenType["relationalOrEqual"] = relationalOrEqual] = "relationalOrEqual"; // <=/>= prec:8
  const bitShiftL = 47113; TokenType[TokenType["bitShiftL"] = bitShiftL] = "bitShiftL"; // << prec:9
  const bitShiftR = 48137; TokenType[TokenType["bitShiftR"] = bitShiftR] = "bitShiftR"; // >>/>>> prec:9
  const plus = 49802; TokenType[TokenType["plus"] = plus] = "plus"; // + prec:10 prefix startsExpr
  const minus = 50826; TokenType[TokenType["minus"] = minus] = "minus"; // - prec:10 prefix startsExpr
  const modulo = 51723; TokenType[TokenType["modulo"] = modulo] = "modulo"; // % prec:11 startsExpr
  const star = 52235; TokenType[TokenType["star"] = star] = "star"; // * prec:11
  const slash = 53259; TokenType[TokenType["slash"] = slash] = "slash"; // / prec:11
  const exponent = 54348; TokenType[TokenType["exponent"] = exponent] = "exponent"; // ** prec:12 rightAssociative
  const jsxName = 55296; TokenType[TokenType["jsxName"] = jsxName] = "jsxName"; // jsxName
  const jsxText = 56320; TokenType[TokenType["jsxText"] = jsxText] = "jsxText"; // jsxText
  const jsxEmptyText = 57344; TokenType[TokenType["jsxEmptyText"] = jsxEmptyText] = "jsxEmptyText"; // jsxEmptyText
  const jsxTagStart = 58880; TokenType[TokenType["jsxTagStart"] = jsxTagStart] = "jsxTagStart"; // jsxTagStart startsExpr
  const jsxTagEnd = 59392; TokenType[TokenType["jsxTagEnd"] = jsxTagEnd] = "jsxTagEnd"; // jsxTagEnd
  const typeParameterStart = 60928; TokenType[TokenType["typeParameterStart"] = typeParameterStart] = "typeParameterStart"; // typeParameterStart startsExpr
  const nonNullAssertion = 61440; TokenType[TokenType["nonNullAssertion"] = nonNullAssertion] = "nonNullAssertion"; // nonNullAssertion
  const _break = 62480; TokenType[TokenType["_break"] = _break] = "_break"; // break keyword
  const _case = 63504; TokenType[TokenType["_case"] = _case] = "_case"; // case keyword
  const _catch = 64528; TokenType[TokenType["_catch"] = _catch] = "_catch"; // catch keyword
  const _continue = 65552; TokenType[TokenType["_continue"] = _continue] = "_continue"; // continue keyword
  const _debugger = 66576; TokenType[TokenType["_debugger"] = _debugger] = "_debugger"; // debugger keyword
  const _default = 67600; TokenType[TokenType["_default"] = _default] = "_default"; // default keyword
  const _do = 68624; TokenType[TokenType["_do"] = _do] = "_do"; // do keyword
  const _else = 69648; TokenType[TokenType["_else"] = _else] = "_else"; // else keyword
  const _finally = 70672; TokenType[TokenType["_finally"] = _finally] = "_finally"; // finally keyword
  const _for = 71696; TokenType[TokenType["_for"] = _for] = "_for"; // for keyword
  const _function = 73232; TokenType[TokenType["_function"] = _function] = "_function"; // function keyword startsExpr
  const _if = 73744; TokenType[TokenType["_if"] = _if] = "_if"; // if keyword
  const _return = 74768; TokenType   De}   N$   t  h   譋    paYPeEE%>        $PheME&8 Ph  E' Yt3E(@ G#  GwPR P:tt(NPheME)8 hPEE*P tZ tK  EPe@    VePNQ9Vet  38E0  <<P  WeVOGWNFOtWettC(%  VeVƆ   etGM&MEt1EE tx+ЃT2  0UQQVE8  P1N  uP  ` NVtMt> ^ UQQVEx  PM  uP  ` NVtMtsMtt4  > ^ UQQVE  PM  uPt7Mt%> ^ j(3W1 M]3E}Ph^D^EVP3s7 M   stFstFM؉u܅t
MEEPE ]PEEP  ̒ x  ` MM}tkMԅt_E    L   j h^Vj Q6    KtAKtAUMtUEMX  3ۋxXMMԉ}t]Mt]`   EPg tZ0  j h^Vj R 6 t3ẺEЉEEM̋]Qp$  p  
 j Y}3E@ GwG΍EP
 M̉u}E 	C  Yj3W/ M3E}Ph`hH]PuE{5 ؃3tc9su^MQphe֋]PEEP   x  ` MM}tE M܅t
]`   EP td/  jh0  4W60 e s j Y3ɉHHj	X  PE   OٕٕٕPٝEۅ  Vٝ ۅٝم ٝحٝeE0   ;  j	^Q0đ    م؅ٝمAteمٝم{Lم؅ٝم{'مٝم{
2tQ   j hahp]j 02 	t  P  PBO Oj hTahp]j Q2 tj P  PE0  蓇
 E b0 =  U%  j hahp]j 0)2     P	  MHEP-ME   P     ƅuƅ tM̉!E   t :/ taT ^ ;Vet^M,  jl4W++ 3ۉ]j] YXX]  E   PMEE   NPzM#RMEREK$   tEPv  L$      EP  }7;tx}E Etn@EEEj@Ee Y聽؍FP]Kn  MEe CEYEE PHQEPMk  6;u]}7Ch5`eF4N0t@F4MEEEPREMt6;učMMM܅tv)  jhU) } tPE8  P)e Et!u6 } t	ue')  } t	ue2`ejhU5) } tPE8  P#)e Et!u' } t	ue(  } t	ue2`e̋U} VtJL$   tAND$PED$u  d$ d$ D$D$D$ D$Pd ^] `ejh  05Wc) tCp      Vlee l  | x x,hexׅ   x|Qxhej hy^h\|^j 0q- t3tPf  pEPc E .+|x,he׋|;xlVeM3(  `ejxx5Wu' EE  EPe u;FuPV? Y2&    PMEuUUU]NEM;   MQ0M\    EM 0PMEEH<ɃQMEEQ]E$<eE$<eEE]E$<eE$<eYu+E+߉}ԍMu؉]MEP|PN  MTEMtEE;EuMtutNV Y`e̋VNtV+у3FFF ^jhUq% u> tSE8  P]%e Et$uVO } t	ue$  } t	ue2`ej V% ]393     PMuEԋMP3uԉu؉u܉uχ tq;MQpheEQ@uP)MtEԍN(FE؉FE܉F ESF$)Mtu!$  Mt_2`ejh  05W.% |CKpt8      Vlee l  x x,he|l  ׅtx|Ql  xhe|l  j hy^h\|^j 0*) |uCl  xx,he|l  ׋x;rVe2M$  P  tEpPa` &Ve`ej
V" EPO "e    EtS0\EuVStWE0P4K  P`} t	ueK"  jh	  5Wo# \C{Kdx|  P  e Ah  ۅXݝh݅hI  G7  ۅ\ݝhܝhA  8PNXd3@E8lP87P   he֋dhWPP03蛱Cݕ`zݝ`\HP  	C ݅`HPxsPQQ$WhP8PP  PR
 ELtNtP3F@Q(%  Etx;8tWj h<ah_j 7%& xt"|Pqhex֋t;8u\|PQ@I  3PTXPE  P T1|xQ8GpheO֋x;PuǍPE9OC @
u5r{A{ƅl ql|4ǅ@; ݝ@ݝ @ݝ(?7U  |q00WVPnU  Dttt8V Ytt<t؍D"M}  `ej]U EPe 8j u   he֋Mtm  UQQE"  Pi8 tMPMt5E `ej
V EP"  e    ]t	P8WuVmt	EWP8F  P   t} t	ue_  j]U pEPe 8u   he֋Mtb  j]U8 !EP7e u8uu7uuNpheVpMtE  j_U EPNX~e 8uup4he֋MtEc  j]U EXP~e 8up<he֋MtoE  UQu	  EY] jh   6W" CNX|tP7~e 	  8P諊EP|r$he֍cxt|M  jh   6W CNX|xtP}e 	  8PEP|r,he֍xt?|M  jh   6W CNX|xtP}e 	  8P苉EP|r0he֍Cxt诿|Mg  j X6W| ـ} ^  C(%  VeVƆ   eEPX  5  e Eԅtƀ  EPK U|E39H$`$ AMMt  WewWetuuWewEWGeE;tP  KM謒  EVe  E;t3+  UECUES+  VeuE;t"MEQHMnLE;uuvEP臓MjZvM؅t3*,  Wewg Wet
j   y  j(6W EMQ  ue F"  @E;tc;MQpheE8pheE MЉEԉU؅trMESP*  ;]uu   @E;t[;MQpheE8pheE MЉEԉU؅tMEP)  ;]uuMM]胻   t"~t#7  N
F 6;uV  `ej6Ww u	  AƇS$   f  T$  
  PV   M"e   Ve  E;tMEB
MEJVe} tnEPT#  EuE;t1]܋KP9Kt薷CQtGMIE;uE } t	ueP#  }b MEpEP辐MjZ譩L#  Rb   jh  6W3 Xced,   x    *  $ u@  *     ݕݕݕPǅ   case TokenType._new:
      return "new";
    case TokenType._this:
      return "this";
    case TokenType._super:
      return "super";
    case TokenType._class:
      return "class";
    case TokenType._extends:
      return "extends";
    case TokenType._export:
      return "export";
    case TokenType._import:
      return "import";
    case TokenType._yield:
      return "yield";
    case TokenType._null:
      return "null";
    case TokenType._true:
      return "true";
    case TokenType._false:
      return "false";
    case TokenType._in:
      return "in";
    case TokenType._instanceof:
      return "instanceof";
    case TokenType._typeof:
      return "typeof";
    case TokenType._void:
      return "void";
    case TokenType._delete:
      return "delete";
    case TokenType._async:
      return "async";
    case TokenType._get:
      return "get";
    case TokenType._set:
      return "set";
    case TokenType._declare:
      return "declare";
    case TokenType._readonly:
      return "readonly";
    case TokenType._abstract:
      return "abstract";
    case TokenType._static:
      return "static";
    case TokenType._public:
      return "public";
    case TokenType._private:
      return "private";
    case TokenType._protected:
      return "protected";
    case TokenType._override:
      return "override";
    case TokenType._as:
      return "as";
    case TokenType._enum:
      return "enum";
    case TokenType._type:
      return "type";
    case TokenType._implements:
      return "implements";
    default:
      return "";
  }
}
